### DISCUSSION TITLE
Bidirectional edges lost converting from (Multi)DiGraph to MultiGraph

### DISCUSSION THREAD

**Discussion by taylorn3** [networkx/networkx#6293] [Q&A]
When I have a directed graph/multigraph that happens to have a bidirectional edge (two nodes that have an edge in each direction between them), I would like them both to be preserved when converting to a MultiGraph.  But it seems that the attributes of one direction are retained, while the other direction's edges are just lost.

An example:

```
mygraph = nx.MultiDiGraph()
mygraph.add_edge("a","b",myattr = "123")
mygraph.add_edge("b","a",myattr = "321")
mygraph2=nx.MultiGraph(mygraph)
```

Now mygraph2 is a MultiGraph that has only one edge connecting a to b, with myattr = "123".  The same thing happens if I use `.to_undirected()`.  Is there a better way to transform a (multi)digraph into a multigraph that keeps all edges?  Of course, I could just fill in all the missing edges with extraneous code, but I would expect there to be an easier solution.  Thanks!

Edit: As it turns out after further testing, if the MultiDiGraph's edges have different keys, then `.to_undirected()` does work.  But there are no keys for a DiGraph, and changing the keys in a MultiDiGraph seems like a bit of a hassle.  But perhaps that's the best solution?

Edit 2: The in-code documentation says the following for the return of to_undirected():

> G : MultiGraph
> An undirected graph with the same name and nodes and
> with edge (u, v, data) if either (u, v, data) or (v, u, data)
> is in the digraph.  If both edges exist in digraph and
> their edge data is different, only one edge is created
> with an arbitrary choice of which edge data to use.
> You must check and correct for this manually if desired.

So I suppose this manual process is the best option, as suggested by the documentation authors.

**Comment by rossbar** [comment]
Note that this behavior isn't specific to MultiGraph, the same holds true for graphs without multiple edges:

```python
>>> G = nx.DiGraph()
>>> G.add_edge(0, 1, color="red")
>>> G.add_edge(1, 0, color="blue")
>>> G.edges(data=True)
OutEdgeDataView([(0, 1, {'color': 'red'}), (1, 0, {'color': 'blue'})])
>>> G.to_undirected().edges(data=True)
EdgeDataView([(0, 1, {'color': 'blue'})])
```

The reason manual intervention is required is that there is no default, obvious way to handle the edge attribute reduction that works for every case. It sounds like your expectation was concatenation so that you'd end up with e.g. `myattr = ["123", "321"]`, which is perfectly valid, but others might want a different operation e.g. sum, min, median, etc. - there's any number of valid reduction operations.

So the current default reduction operation is essentially "arbitrary sampling"[^1], and users who need different behavior have to implement it manually. At least the documentation makes this clear-ish (small victories!).

[^1]: arbitrary != random

**Reply by taylorn3** [reply]
Thanks for the response!  I guess what I meant was that I wanted to have parallel (undirected) edges, each edge from the original (multi)digraph turned into an edge.  So there isn't an issue with the attributes (as they will be part of their own edges), but perhaps the same issue applies when deciding the keys?

In any case, I was able to write a shorter amount of code than I perhaps thought, which solved the issue by giving every edge its own unique key (here mygraph is actually a DiGraph, probably could be fixed in a similar way for MultiDiGraphs):
```
mygraph2 = nx.MultiDiGraph()
mygraph2.add_nodes_from(mygraph.nodes(data=True))
mygraph2.add_edges_from([(u,v,w,x) for w,(u,v,x) in enumerate(mygraph.edges(data=True))])
mygraph2 = mygraph2.to_undirected()
```
